token-upgrade/program/src/instruction.rs

```diff
@@ -0,0 +1,48 @@
+//! Program instructions
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// Instructions supported by the TokenUpgrade program
+#[derive(Clone, Copy, Debug, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
+pub enum TokenUpgradeInstruction {
+    /// Creates an upgrade factory for the given mint.
+    ///
+    /// Returns an error if the account exists.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writeable,signer]` Funding account (must be a system account)
+    ///   1. `[writeable]` PDA factory account to be created
```

Comment:
Maybe this account can be avoided entirely since it only seems to hold `pre_minted_token_account`, which is already being passed into `UpdateTokens`?

Reply:
That's a good point, maybe the entire thing is stateless, and literally just does a burn on one side, and a transfer on the other. We simply define a standard for how to derive the owner of the pre-minted token account, as `PDA(old_mint, new_mint)`, so that way you can always find the accounts to use as long as you know the two mint addresses.

The entire idea of the factory was to give a pointer from the old mint to the new mint, so you can easily discover the new mint.  And to avoid scams, every mint only has one factory, so there's only one possible upgrade path. Since there's only one factory, we force the mint authority to sign to avoid address squatting.

But the more we talk about this, the more I realize that this is trying too hard, and we're better off making it entirely stateless.  What do you think?

Reply:
Yep! Glad we arrived at the same place independently.